noir-projects/noir-protocol-circuits/crates/private-kernel-lib/src/private_call_data_validator.nr

```diff
@@ -0,0 +1,225 @@
+    nullifier_key_validation_requests: u64,
+    new_note_hashes: u64,
+    new_nullifiers: u64,
+    new_l2_to_l1_msgs: u64,
+    private_call_stack_hashes: u64,
+    public_call_stack_hashes: u64,
+    encrypted_logs_hashes: u64,
+    unencrypted_logs_hashes: u64,
+}
+
+struct PrivateCallDataValidator {
+    data: PrivateCallData,
+    array_lengths: ArrayLengths,
+}
+
+impl PrivateCallDataValidator {
+    pub fn new(data: PrivateCallData) -> Self {
+        let array_lengths = validate_arrays(data);
+        PrivateCallDataValidator { data, array_lengths }
+    }
+
+    pub fn validate(self) {
+        self.validate_contract_address();
+        self.validate_call();
+        self.validate_private_call_requests();
+        self.validate_public_call_requests();
+        self.validate_teardown_call_request();
+    }
+
+    // Confirm that the TxRequest (user's intent) matches the private call being executed.
+    pub fn validate_against_tx_request(self, tx_request: TxRequest) {
+        let call_stack_item = self.data.call_stack_item;
+        assert_eq(
+            tx_request.origin, call_stack_item.contract_address, "origin address does not match call stack items contract address"
+        );
+        assert_eq(
+            tx_request.function_data.hash(), call_stack_item.function_data.hash(), "tx_request function_data must match call_stack_item function_data"
+        );
+        assert_eq(
+            tx_request.args_hash, call_stack_item.public_inputs.args_hash, "noir function args passed to tx_request must match args in the call_stack_item"
+        );
+        assert_eq(
+            tx_request.tx_context, call_stack_item.public_inputs.tx_context, "tx_context in tx_request must match tx_context in call_stack_item"
+        );
+    }
+
+    pub fn validate_against_call_request(self, request: CallRequest) {
+        let call_stack_item = self.data.call_stack_item;
+
+        assert_eq(
+            request.hash, call_stack_item.hash(), "calculated private_call_hash does not match provided private_call_hash at the top of the call stack"
+        );
+
+        let call_context = call_stack_item.public_inputs.call_context;
+
+        if call_context.is_delegate_call {
+            let caller_context = request.caller_context;
+            assert(!caller_context.is_empty(), "caller context cannot be empty for delegate calls");
+            assert_eq(
+                call_context.msg_sender, caller_context.msg_sender, "call stack msg_sender does not match expected msg_sender for delegate calls"
+            );
+            assert_eq(
+                call_context.storage_contract_address, caller_context.storage_contract_address, "call stack storage address does not match expected contract address for delegate calls"
+            );
+        } else {
+            assert_eq(
+                call_context.msg_sender, request.caller_contract_address, "call stack msg_sender does not match caller contract address"
+            );
```

Comment:
Moved the checks for `storage_contract_address` vs `contract_address` to `validate_call()` in this file. Since they don't involve a call request.